c2rust-analyze/src/rewrite/expr/mir_op.rs

```diff
@@ -1404,6 +1515,14 @@ pub struct CastBuilder<'a, 'tcx, PT1, PT2, F> {
     perms: &'a PT1,
     flags: &'a PT2,
     emit: F,
+    /// If set, the cast builder is allowed to move the expression being cast.  For example, we
+    /// might emit `ptr.map(|p| &mut *p).unwrap()` instead of `ptr.as_deref_mut().unwrap()` when
+    /// `can_move` is set; the former moves out of `ptr` but avoids an additional borrow, making it
+    /// suitable to be used as the result expression of a function.
+    can_move: bool,
+    /// If set, the cast builder will emit a downgrade/borrow operation even for no-op casts, if
+    /// the thing being cast can't be moved (`!can_move`) and also can't be copied.
```

Comment:
Are there only 3 states between the 2 variables: can-move, can-copy, must-borrow? Maybe using an `enum` here would be better.

Reply:
There are indeed only three possible behaviors - `can_move && borrow` produces the same cast as `can_move && !borrow`.  However, the code that sets the flags handles the two independently (and in particular, sometimes produces the `can_move && borrow` case).  So I think it's better to leave the code as-is, with the logic mapping the four cases to three outcomes centralized in the builder.